From 98518848a284edee58305a203365d492bf10f45b Mon Sep 17 00:00:00 2001
From: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
Date: Wed, 27 Dec 2023 12:21:07 +0530
Subject: [PATCH] ath11k: Change profile count to bssid indicator

When VDEV up, the host sends the profile count to the firmware
as a non-transmitting VAP count. According to firmware request,
profile count is expected at 2 power from max bssid_indicator.

Therefore, we update the profile count with 2 power of
maximum BSSID indicator.

Signed-off-by: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/mac.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -1821,8 +1821,6 @@ static void ath11k_control_beaconing(str
 				    arvif->vdev_id, ret);
 
 		arvif->is_up = false;
-		if (tx_arvif)
-			tx_arvif->nontransmitting_vif_count = 0;
 
 		return;
 	}
@@ -1850,9 +1848,7 @@ static void ath11k_control_beaconing(str
 	if (tx_arvif) {
 		params.tx_bssid = tx_arvif->bssid;
 		params.profile_idx = info->bssid_index;
-		if (params.profile_idx >= tx_arvif->nontransmitting_vif_count)
-			tx_arvif->nontransmitting_vif_count = params.profile_idx;
-		params.profile_count = tx_arvif->nontransmitting_vif_count;
+		params.profile_count = BIT(info->bssid_indicator);
 	}
 	ret = ath11k_wmi_vdev_up(arvif->ar, &params);
 	if (ret) {
@@ -3409,8 +3405,6 @@ static void ath11k_bss_disassoc(struct i
 	arvif->is_up = false;
 	if (arvif->vif->bss_conf.mbssid_tx_vif) {
 		tx_arvif = (void *)arvif->vif->bss_conf.mbssid_tx_vif->drv_priv;
-		if (tx_arvif != arvif)
-			tx_arvif->nontransmitting_vif_count--;
 	}
 
 	memset(&arvif->rekey_data, 0, sizeof(arvif->rekey_data));
@@ -9438,7 +9432,7 @@ ath11k_mac_update_vif_chan(struct ath11k
 			tx_arvif = (void *)arvif->vif->bss_conf.mbssid_tx_vif->drv_priv;
 			params.tx_bssid = tx_arvif->bssid;
 			params.profile_idx = arvif->vif->bss_conf.bssid_index;
-			params.profile_count = tx_arvif->nontransmitting_vif_count;
+			params.profile_count = BIT(arvif->vif->bss_conf.bssid_indicator);
 		}
 		ret = ath11k_wmi_vdev_up(arvif->ar, &params);
 		if (ret) {
--- a/drivers/net/wireless/ath/ath11k/core.h
+++ b/drivers/net/wireless/ath/ath11k/core.h
@@ -523,7 +523,6 @@ struct ath11k_vif {
 	struct ath11k_tid_qos_config tid_cfg[ATH11K_TID_MAX];
 	u32 tids_rst;
 	u64 tbtt_offset;
-	u8 nontransmitting_vif_count;
 	struct work_struct update_bcn_template_work;
 	DECLARE_BITMAP(free_groupidx_map, ATH11K_GROUP_KEYS_NUM_MAX);
 };
